### PR TITLE
[WIP] fix checkpoint hook logic

### PR DESCRIPTION
**Pull request by BIGWangYuDong**
## Motivation

If use EpochBasedRunner and save ckpt as IterBase (such as `checkpoint_config = dict(by_epoch=False, interval=100)`), this may cause a potential error: the ckpt will name as `epoch_*.pth`.

## Modification

change the logic in checkpoint_hook.

 Maybe we can only set the code in BaseRunner?

### CODE REVIEW

File: mmcv/runner/epoch_based_runner.py [hunk 1]
```diff
@@ -134,7 +134,8 @@ def save_checkpoint(self,
                         filename_tmpl='epoch_{}.pth',
                         save_optimizer=True,
                         meta=None,
-                        create_symlink=True):
+                        create_symlink=True,
+                        by_epoch=True):
```
Comment (AllentDan) [inline mmcv/runner/epoch_based_runner.py:138]: Lack corresponding docstring here.

File: mmcv/runner/iter_based_runner.py [hunk 2]
```diff
@@ -181,7 +181,8 @@ def save_checkpoint(self,
                         filename_tmpl='iter_{}.pth',
                         meta=None,
                         save_optimizer=True,
-                        create_symlink=True):
+                        create_symlink=True,
+                        by_epoch=False):
```
Comment (AllentDan) [inline mmcv/runner/iter_based_runner.py:185]: Lack docstring

**Review by AllentDan — COMMENTED** [review]
May add unit tests

**Comment by BIGWangYuDong** [comment]
- Move the code to `BaseRunner` and delete the `save_checkpoint` function in `EpochBasedRunner` and `IterBasedRunner`
- update related test unit
- add a warning if the checkpoint filename template does not match the saving checkpoints way

File: mmcv/runner/base_runner.py [hunk 3]
```diff
@@ -195,14 +197,70 @@ def val(self):
     def run(self, data_loaders, workflow, **kwargs):
         pass
 
-    @abstractmethod
     def save_checkpoint(self,
                         out_dir,
-                        filename_tmpl,
+                        filename_tmpl='epoch_{}.pth',
```
Comment (AllentDan) [inline mmcv/runner/base_runner.py:202]: Using `epoch_{}.pth` as the default checkpoint name for `iter_based_runner` may be a little confusing for users. A possible method is using None as default, but getting changed to `iter_{}.pth` or `epoch_{}.pth` according to `by_epoch` if it is None. But I am not sure if it is the best solution here.

**Pull request closed without merging**